### ISSUE TITLE
Interoperability issues with dependencies using Objective-C in Cluster Mode on macOS 10.13 High Sierra

### ISSUE THREAD

**Issue by ticky** [puma/puma#1421]
### Steps to reproduce

1) Install macOS 10.13 High Sierra

2) Freshly install Ruby and Puma

3) Execute an application which uses the `pg` gem, does not use `preload_app!`, and uses Puma in cluster mode

### Expected behavior

Puma should work and execute jobs as expected

### Actual behavior

Puma fails, repeatedly outputting these errors from worker processes:

```
objc[81924]: +[__NSPlaceholderDictionary initialize] may have been in progress in another thread when fork() was called.
objc[81924]: +[__NSPlaceholderDictionary initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
```

### System configuration

**Ruby version**: 2.4.2
**Rails version**: 5.1.3
**Puma version**: 3.10.0

macOS High Sierra [changes the behaviour of the `fork` syscall](http://sealiesoftware.com/blog/archive/2017/6/5/Objective-C_and_fork_in_macOS_1013.html) such that calls to Objective-C APIs in forked processes are treated as errors.

~It appears some part of Puma’s use of Ruby violates this, which means Puma doesn’t work correctly in cluster mode on 10.13.~

~A temporary workaround is to apply the `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` environment variable to the parent puma process, which will then trigger the behaviour observable in macOS versions 10.12 and earlier.~

EDIT: Puma uses `fork` in cluster mode, and no special consideration is given to what runs within the fork. It is possible to depend upon a gem, or dynamically load a shared object, which in turn links to an Objective-C framework. This linkage will implicitly call an Objective-C `initialize` method as it sets up the Objective-C runtime for the first time, and if this call occurs within a fork, the program will be killed, and the above error message printed.

A less heavy-handed workaround is to use a `before_fork` callback like this:

```ruby
# Work around macOS 10.13 and later being very picky about
# `fork` usage and interactions with Objective-C code
# see: <https://github.com/puma/puma/issues/1421>
if /darwin/ =~ RUBY_PLATFORM
  before_fork do
    require 'fiddle'
    # Dynamically load Foundation.framework, ~implicitly~ initialising
    # the Objective-C runtime before any forking happens in Puma
    Fiddle.dlopen '/System/Library/Frameworks/Foundation.framework/Foundation'
  end
end
```

This will implicitly initialise the Objective-C framework within the Puma process, allowing forked children to use other Objective-C frameworks.

**Comment by nateberkopec** [comment]
Relevant Unicorn thread https://bogomips.org/unicorn-public/20170804191023.GA28511@dcvr/T/

**Comment by nateberkopec** [comment]
I'm with Eric. I have no idea how ObjC factors into this.

**Comment by ticky** [comment]
ObjC factors in because if anything within the forked process links against ObjC APIs, it will be unusable from within Puma! :)

I’m doing some more debugging to determine what exactly is causing this.

**Comment by mperham** [comment]
Oh crap. Subscribe me, github!

**Comment by ticky** [comment]
Having trouble doing more debugging, as the child exits as soon as it encounters this error, so attaching rbtrace seems difficult if not impossible.

**Comment by boazsegev** [comment]
Sorry to butt in, but do you know if this effects [`iodine`](https://github.com/boazsegev/iodine) as well?

I can't install High Sierra since many of the professional applications I use daily don't support it just yet (I'm also a musician).

**Comment by ticky** [comment]
Anything that forks and runs user-specified code within that fork without using `exec` may be affected by this.

It is unclear exactly where a solution will need to be implemented.

**Comment by ticky** [comment]
I’ve managed to track down what appears to be the source of this error in the application I’m working with, it’s loading [`pg`](https://rubygems.org/gems/pg) which has a native extension linked against Postgres’ `libpq.5` which in turn is linked against Kerberos.framework and LDAP.framework.

It seems unlikely that we can get away with our workers not being able to talk to our database on macOS… 🤔

Again, this is not that you have to actively be _using_ any features provided by these Objective-C frameworks, their _inclusion in a thing you link against_ is enough to trigger this error. We do not use the Kerberos nor LDAP features of Postgres in our application, and yet here we have this error.

**Comment by ticky** [comment]
Apparently this thread has [ended up on reddit](https://www.reddit.com/r/ruby/comments/72qxjv/kernelfork_broken_in_macos_1013_high_sierra_puma/) where it seems to be being misunderstood.

Regarding [this comment](https://www.reddit.com/r/ruby/comments/72qxjv/kernelfork_broken_in_macos_1013_high_sierra_puma/dnl8unj/)

>I'm trying to come up with a summary of the issue; please assist:
>1. macOS includes an Objective-C runtime
>2. macOS includes an OS framework based on the Objective-C runtime
>3. Objective-C classes defined by the OS framework are fork-unsafe, meaning that fork() callers should not execute code before calling exec()
>4. macOS 10.13 includes a macOS 10.13 SDK (presumably based on the OS framework but not responsible for defining the OS framework's classes)
>5. macOS applications built with the macOS 10.13 SDK are able to use new SDK features in order to execute Objective-C code between fork() and exec(), therefore achieving the fork-safe property when used correctly
>6. There remains a tension between fork-safety and thread-safety, due to the structure of the locks held by the Objective-C runtime
>7. As such, there are restrictions in the mac OS 10.13 SDK about the nature of calls to +initialize between calling fork() and exec()
>8. Thus, the new capabilities in the mac OS 10.13 SDK for calling code safely between fork() and exec() are guarded by protections which enforce the limitations to +initialize in order to maintain fork-safe and thread-safe properties
>
> What is the connection to Ruby on macOS 10.13? I'm not very familiar with this platform, but it would appear that the affected Ruby libs on macOS are making use of the SDK, calling fork(), and running into the new behavior.

The issue is that code within Ruby can call fork (as Puma does), and loading dynamic libraries from Ruby or a Ruby native extension can cause an Objective-C class to be `+initialize`d, triggering this error message and a swift exit.

**Comment by boazsegev** [comment]
@ticky ,

I'm happy you're managing to track the issue down and I'm thankful for the added information.

However, reading the information it wasn't clear to me if `pg` might be producing the error independently of the server (does `pg` use `fork`?) and  whether or not `pg`'s initialization could be handled pre-forking (perhaps avoiding the issue)?

1. Does the error occur also in single process mode (in `puma` and `iodine`, add the `-w 1` to the command line)?

2. Could you initialize the`pg` database within the `config.ru` and see if pre-fork initialization might work around the issue? (remember to re-establish the DB connections, so they aren't shared acrid processes)...?

Just my 2¢, I don't know if any of it helps.

P.S.

You posted the issue summery as I was writing this and I wondered if, perhaps, some non Objective-C function calls are implemented using the Objective-C framework under the hood..?

**Comment by ticky** [comment]
> wasn't clear to me if `pg` might be producing the error independently of the server (does `pg` use fork?) 

I don’t believe `pg` uses `fork` itself. These errors are produced by workers immediately after spinning up. It appears to occur entirely within the initialisation of our application's dependencies.

> whether or not `pg`'s initialization could be handled pre-forking

This is the best hypothesis I now have. It looks like it _might_ work if those things are all loaded before forking, however, I am not familiar enough with the way our system is set up to know how the should work!

> You posted the issue summery as I was writing this and I wondered if, perhaps, some non Objective-C function calls are implemented using the Objective-C framework under the hood..?

I don’t believe it’s that they are intentionally calling out to any particular Obj-C method, but merely linking to their Obj-C frameworks which requires the Obj-C initialisation to occur!

**Comment by mistydemeo** [comment]
> However, reading the information it wasn't clear to me if pg might be producing the error independently of the server (does pg use fork?) and whether or not pg's initialization could be handled pre-forking (perhaps avoiding the issue)?

`pg` doesn't use `fork`, but the issue isn't active initialization by `pg` so much as how the ObjC runtime startup is occurring. In this context since the program itself isn't written in ObjC and doesn't directly have any ObjC linkage, the ObjC runtime only enters the equation when a `dlopen` call (via Ruby's `dln_load`) loads a shared object which has ObjC linkage. As you can see in this truncated backtrace from an aborted Ruby process the dynamic loader is initializing the ObjC runtime automatically, which causes CoreFoundation to initialize an object. That triggers the kill.

```
* thread #1, stop reason = signal SIGSTOP
  * frame #0: 0x00007fff53bde336 libsystem_kernel.dylib`__abort_with_payload + 10
    frame #1: 0x00007fff53bd8e00 libsystem_kernel.dylib`abort_with_payload_wrapper_internal + 89
    frame #2: 0x00007fff53bd8da7 libsystem_kernel.dylib`abort_with_reason + 22
    frame #3: 0x00007fff52ea2962 libobjc.A.dylib`_objc_fatalv(unsigned long long, unsigned long long, char const*, __va_list_tag*) + 108
    frame #4: 0x00007fff52ea2814 libobjc.A.dylib`_objc_fatal(char const*, ...) + 135
    frame #5: 0x00007fff52ea343b libobjc.A.dylib`performForkChildInitialize(objc_class*, objc_class*) + 341
    frame #6: 0x00007fff52e93e8f libobjc.A.dylib`lookUpImpOrForward + 228
    frame #7: 0x00007fff52e93914 libobjc.A.dylib`_objc_msgSend_uncached + 68
    frame #8: 0x00007fff52e97145 libobjc.A.dylib`+[NSObject new] + 86
    frame #9: 0x00007fff2e5bb025 Foundation`-[NSThread init] + 61
    frame #10: 0x00007fff2e5bf0f5 Foundation`____NSThreads_block_invoke + 64
    frame #11: 0x00007fff53a55f64 libdispatch.dylib`_dispatch_client_callout + 8
    frame #12: 0x00007fff53a55f17 libdispatch.dylib`dispatch_once_f + 41
    frame #13: 0x00007fff2e5bafe3 Foundation`_NSThreadGet0 + 325
    frame #14: 0x00007fff2e5ba7ef Foundation`_NSInitializePlatform + 341
    frame #15: 0x00007fff52e958df libobjc.A.dylib`call_load_methods + 239
    frame #16: 0x00007fff52e92c16 libobjc.A.dylib`load_images + 70
    frame #17: 0x000000011125d198 dyld`dyld::notifySingle(dyld_image_states, ImageLoader const*, ImageLoader::InitializerTimingList*) + 407
    frame #18: 0x000000011126d15b dyld`ImageLoader::recursiveInitialization(ImageLoader::LinkContext const&, unsigned int, char const*, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 309
    frame #19: 0x000000011126d103 dyld`ImageLoader::recursiveInitialization(ImageLoader::LinkContext const&, unsigned int, char const*, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 221
    frame #20: 0x000000011126d103 dyld`ImageLoader::recursiveInitialization(ImageLoader::LinkContext const&, unsigned int, char const*, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 221
    frame #21: 0x000000011126d103 dyld`ImageLoader::recursiveInitialization(ImageLoader::LinkContext const&, unsigned int, char const*, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 221
    frame #22: 0x000000011126d103 dyld`ImageLoader::recursiveInitialization(ImageLoader::LinkContext const&, unsigned int, char const*, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 221
    frame #23: 0x000000011126c2a6 dyld`ImageLoader::processInitializers(ImageLoader::LinkContext const&, unsigned int, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 134
    frame #24: 0x000000011126c33a dyld`ImageLoader::runInitializers(ImageLoader::LinkContext const&, ImageLoader::InitializerTimingList&) + 74
    frame #25: 0x00000001112603e5 dyld`dyld::runInitializers(ImageLoader*) + 82
    frame #26: 0x0000000111269002 dyld`dlopen + 527
    frame #27: 0x00007fff53a90eb6 libdyld.dylib`dlopen + 86
    frame #28: 0x0000000102d0ce3f ruby`dln_load + 175
```

**Comment by ticky** [comment]
> in single process mode

I had tried to activate single-process mode earlier but not had any luck. Turns out you need to remove the `workers` configuration statements altogether!

Running in single mode (confirmed with `Puma starting in single mode…` output!) works as expected with no crashes or errors.

Thus, it’s the `fork` by Puma (or whatever else is using it), not anything within `pg` which is causing the ruckus. 😄

**Comment by ticky** [comment]
A bit of an update - this behaviour also doesn’t cause a crash if `preload_app!` is set to true. I can’t seem to find the implementation of this feature, though?

**Comment by mistydemeo** [comment]
I have a Unicorn app, rather than a Puma app, but something that's working for me is to explicitly require a gem with ObjC linkage within a `before_fork` block. That ensures that the ObjC runtime is loaded and initialized before the fork occurs, and it's a bit lighter-weight than `preload_app!`.

**Comment by ticky** [comment]
@mistydemeo’s onto something! I just tried adding `require 'pg'` to our Puma `before_fork`, and the errors have ceased! Thanks so much!

I wonder if there’s a less strange-looking way to achieve this than loading a specific dependency we know to link ObjC. 🤔

**Comment by ticky** [comment]
Okay, this doesn’t look much _less_ strange but it is, at least, only relying on side effects it’s explicitly calling upon!

```ruby
# Work around macOS 10.13 and later being very picky about
# `fork` usage and interactions with Objective-C code
# see: <https://github.com/puma/puma/issues/1421>
if /darwin/ =~ RUBY_PLATFORM
  before_fork do
    require 'fiddle'
    # Dynamically load Foundation.framework, ~implicitly~ initialising
    # the Objective-C runtime before any forking happens in Puma
    Fiddle.dlopen '/System/Library/Frameworks/Foundation.framework/Foundation'
  end
end
```

As Misty noted, this ensures the the Objective-C runtime has been initialised (a side effect instigated by the dynamic loader itself) before any `fork` happens, which means once `pg` is required and loads the Kerberos and LDAP frameworks (which in turn rely on Foundation), they do not need to perform that implicit `initialize`.

Whether this is a workaround which makes sense to include in Puma is definitely debatable, but it appears that many things which make use of this sort of fork model will need to consider it, and perhaps also document it.

**Comment by ticky** [comment]
Updated the initial post with the updated details of this issue!

**Comment by nateberkopec** [comment]
Great work @ticky! I won't be able to take a closer look at this until next week - holding off on my own High Sierra upgrade until I am back home and can make some backups first.

**Comment by tenderlove** [comment]
According to [this article](http://sealiesoftware.com/blog/archive/2017/6/5/Objective-C_and_fork_in_macOS_1013.html), setting `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` environment variable should work too.  It's really not clear how to fix this correctly (besides not lazily loading things I guess).

**Comment by boazsegev** [comment]
> doesn’t solve the fact that it may not even be possible in future OS versions

It would be possible to have a quite exception instead of the `before_fork` - `fork` could be assumed, in the interest of  C extensions such as `passenger` and `iodine`, where `before_fork` callbacks might not get invoked due to a direct system call (rather than `Kernel.fork`)...

Maybe something like:

# Work around macOS 10.13 and later being very picky about
# `fork` usage and interactions with Objective-C code
# see: <https://github.com/puma/puma/issues/1421>
if /darwin/ =~ RUBY_PLATFORM
  begin
    require 'fiddle'
    # Dynamically load Foundation.framework, ~implicitly~ initialising
    # the Objective-C runtime before any forking happens in Puma
    Fiddle.dlopen '/System/Library/Frameworks/Foundation.framework/Foundation'
  rescue Exception => e
	
  end
end

> it may have other consequences

Since `fiddle` is part of the standard library, it seems safe enough.

However, the `_NS*` bindings might be another issue (introducing name space concerns and possible conflicts), which make this a very valid warning.

**Comment by ticky** [comment]
@boazsegev my last message was regarding the use of the environment variable (`OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES`) - my Ruby-based workaround will likely remain viable. ☺️

As @mistydemeo suggested, it works in Unicorn too, and it likely works in any other thing which is using fork in a similar way, and running arbitrary Ruby code within.

I would suggest that having an option to implicitly load Foundation would make some sense, however, I would be reluctant to suggest having it always implicitly do this for all users, as it feels very heavy-handed. It’s unfortunate that we can’t catch these kinds of errors in our code (the process is killed instantly), or I’d suggest adding our own message about it!

It’s also worth noting again that this behaviour doesn’t trigger this error case outside of cluster mode, or if the application is preloaded.

**Comment by boazsegev** [comment]
@ticky , you're probably right about it being heavy-handed, but...

> it likely works in any other thing which is using `fork` in a similar way...

I don't think it does. Especially since `before_fork` is [puma's DSL](https://github.com/puma/puma/blob/d31d68a610ef132c1e0bd801dd210fc42b0c67b6/lib/puma/dsl.rb#L293-L305) and not a general Ruby feature.

Other C extensions (such as `passenger` and `iodine`) might use a different DSL before they [call the system call directly](https://github.com/phusion/passenger/blob/7eb5f0e9da983fb034a77f02a3a7b21708fd29ae/src/agent/SpawnPreparer/SpawnPreparerMain.cpp#L121-L135) (not using Ruby's `fork`), assuming any required actions were directly performed (rather than placed in a callback).

I might be over thinking this, since the issue was opened here, but the discussion feels as if it's more than about a Puma specific solution (it includes Unicorn, Sidekiq etc').

Just my 2¢, of course.

**Comment by ticky** [comment]
I don’t mean the `before_fork` call specifically, just that you’ll need _some_ mechanism to do the `dlopen` call before you fork. Indeed, its use will depend on what you’re using, and could simply be to put it in whatever calls out to to Passenger or Iodine - it merely has to be done within the same process.

You are right, though, that the issue is bigger than Puma. It’s likely that Puma, Unicorn, Passenger and Iodine will all need to consider the implications of this macOS change. Unicorn’s maintainer, though, seems to have already expressed a disinterest in doing so!

Either that or it requires some sort of structural change in Ruby (though that seems rather unlikely to fly)

**Comment by FooBarWidget** [comment]
It looks like the dlopen workaround is the best solution in terms of fewest side-effects. `OBJC_DISABLE_INITIALIZE_FORK_SAFETY ` is too broad and affects child processes.

Regarding @ticky's comment about structural changes in Ruby: I think this goes beyond Ruby itself. Forking is fundamentally problematic when there is multithreading involved. We can make it work, and in the context of Ruby app servers it *usually* works well so far, but that's in spite of a lack of real solutions.

To *guarantee* that forking is safe, the application must not be running any threads at the point of fork (or the other threads must only be executing async-signal-safe code). MRI can guarantee this about its own code to some degree, but third party native libraries may spawn their own threads. So any gems that the user pulls in should not spawn threads until the app server has forked. Gems with multithreaded native extensions should provide an explicit thread initialization method, which the app only calls after the app server has forked. This requires cooperation in the entire gem ecosystem.

Does something like the dlopen workaround belong in Puma/Unicorn/iodine/Passenger? This is more of a philosophical question than a technical one. I cannot answer for Puma/Unicorn/iodine, but Passenger's philosophy is put user experience at the forefront, so we have chosen to include the dlopen workaround in the next Passenger version.

**Comment by ticky** [comment]
Unless it has other side effects, I’d suggest implementing this within Puma, Unicorn and iodine would cause the fewest headaches for implementors and users.

Barring some incompatibility we haven’t yet encountered with initialising the Objective-C runtime within our Ruby process, I think it makes the most sense!

**Comment by mperham** [comment]
Is there some reason why this shouldn't be passed up to MRI instead of fixed in each individual project?  For instance, MRI on OSX could touch the Obj-C runtime on startup and all this goes away, right?

**Comment by ticky** [comment]
I believe the argument against that would be that things like Puma, Unicorn, iodine and Passenger actually _know_ they will be `fork`ing and never `exec`’ing, whilst Ruby itself doesn’t necessarily know that. I don’t think it would be too detrimental, though. It may even make sense for Ruby on macOS to initialise the Objective-C runtime the first time `fork` is called?

**Comment by boazsegev** [comment]
@mperham, I would generally agree that this seems like a Ruby concern rather than a specific gem (or a group of gems).

I would assume that Ruby (as a scripting language) should behave consistently across supposed environments without requiring code changes (except where different behaviors are desirable). For this reason, an OS specific workaround that's required for a consistent behavior across OS boundaries, should probably (IMHO) be part of the core Ruby code. 

Having said that, I wonder if pushing this into a Ruby release is even an option.

@ticky, In general I find your approach very wise and balanced.

However, I doubt that it is safe to assume that the Ruby version of `fork` would be called whenever Ruby is running in the background.

Processes can be forked by C/Java extensions as well (using direct system calls). Besides, Ruby MRI is also available as an engine/library (much like V8 for Javascript) and could be executed from within an already running environment.

I wonder what price we pay for loading the Objective-C runtime, but it might be safer to preemptively load the Objective-C runtime  during the Ruby VM initialization procedure (although it's somewhat high handed, as you pointed out).

**Comment by FooBarWidget** [comment]
I like the proposal of adding a modification to Ruby.

It is true that native libraries can call fork without going through the Ruby fork function. Having said that, I believe that such libraries already ensure that they perform async signal safe actions inside the child. And if they don't then I believe that it is up to them to fix that, or up to the  wrapping gem to work around that, because those libraries would already cause problems even outside the context of Ruby. 

It is also true that Ruby can be embedded in another environment. But Ruby forking in such environment is at present already a problem. I believe it is up to the embedder to add mechanisms to make Ruby's fork call safe, for example with pthread_atfork.

So I support the notion of having Ruby initialize the objc runtime upon calling #fork.

**Comment by ticky** [comment]
> Processes can be forked by C/Java extensions as well (using direct system calls).

That’s actually a good point, which I hadn’t considered!

> So I support the notion of having Ruby initialize the objc runtime upon calling #fork.

I have just opened [an issue on Ruby’s tracker](https://bugs.ruby-lang.org/issues/14009) for this. I hope we can work something nice out!

**Comment by FooBarWidget** [comment]
I've published [a blog post](https://blog.phusion.nl/2017/10/13/why-ruby-app-servers-break-on-macos-high-sierra-and-what-can-be-done-about-it/) on this subject. This blog post explains:

 - What the problem is and why it occurs
 - The current community concensus on what to do about this
 - What users can expect in the short term from various parties such as app servers authors
 - What they can do themselves

Hopefully this post brings more clarity to users.

**Comment by nateberkopec** [comment]
Great post @FooBarWidget. Just to clarify some things from Puma's perspective:

* I usually side with the user-experience side as well, and since Puma is the default for Rails I think it's *extra* important that we Just Work. I'm generally just watching and waiting this thread and the Ruby issue tracker to see what consensus emerges. If it looks like Ruby will quickly implement a fix, we may not do anything in Puma. But we've done some wild workarounds for Ruby bugs in the past too so I'm not against merging something to fix this here.
* Since v2.1, Puma [warns the user if it detects additional Ruby threads created during preloading](https://github.com/puma/puma/commit/a2288fd6564ee9cdfa2d6596c488114ef2d000c9).

**Comment by ticky** [comment]
An update from the Ruby tracker, [a patch for Ruby has been suggested](https://bugs.ruby-lang.org/issues/14009#note-4) and I’ve confirmed it mitigates this issue when it comes to Puma!

It works around it by linking Ruby itself to Foundation instead of CoreFoundation. CoreFoundation doesn’t initialise Objective-C, but as we know from this thread, Foundation does. 😄

**Comment by FooBarWidget** [comment]
@nateberkopec I think fixing it in Ruby only affects versions of Ruby from a specific version onward. What about users on older versions? Would it be reasonable to ask them to upgrade Ruby or would it be better if you activate the workaround on older Ruby versions?

> Since v2.1, Puma warns the user if it detects additional Ruby threads created during preloading.

This is... an awesome idea! This will solve many of the mysterious issues people have with preforking.

**Comment by ticky** [comment]
> fixing it in Ruby only affects versions of Ruby from a specific version onward

That’s true, however, it seems like it’s both likely to be a thing worthy of back porting to any existing supported versions, and a very simple patch to apply if you need support for an even older version.

**Comment by boazsegev** [comment]
> Would it be reasonable to ask them to upgrade Ruby or would it be better if you activate the workaround on older Ruby versions?

IMHO, since the issue effects High Sierra, I believe it is acceptable to expect users to upgrade to a High Sierra compatible Ruby version.

I also think that back porting the patch is wise.

**Comment by nateberkopec** [comment]
Yeah, I agree with what @ticky and @boazsegev said. Also, since there *is* an environment variable setting which gives you "the old behavior", it's easier for us to just point users to that instead.

IMO "upgrade your ruby patchlevel" or "use this environment variable to go back to the way things were pre-2017" is an OK solution.

**Comment by boazsegev** [comment]
UPDATE: It seems a patch was merged with the Ruby trunk. This probably means the issue will be resolved in future Ruby releases (the next 2.4.x version should be patched).

**Comment by nateberkopec** [comment]
Okay, because of the quick fix [here](https://bugs.ruby-lang.org/issues/14009#note-4) I'm going to close this from our perspective. This will be fixed in Ruby 2.4.3 (which will *probably* be around Christmas), as of this writing backports to other Ruby minor versions are unknown.

To work around this issue *today*, I suggest just setting the `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` environment variable when working with Puma. This gets you the old, "unsafe" behavior. To understand why this is an imperfect solution, [read @FooBarWidget's blog](https://blog.phusion.nl/2017/10/13/why-ruby-app-servers-break-on-macos-high-sierra-and-what-can-be-done-about-it/). Try not to set it permanently or in your .bash_profile or anything, it's a genuinely useful feature that you should probably be using - instead maybe add it to a Procfile or something like that.

**Comment by boazsegev** [comment]
As an update from the iodine server perspective:

The iodine server now includes a fix for this issue, allowing people using `iodine` today to have a better experience.

The decision was made for two reasons:

1. This allows High Sierra to be supported without requiring the Ruby runtime to be upgraded to a High Sierra compatible version.

1. Waiting for Christmas (when I assume the next Ruby version will be released) seemed a bit of a wait...

I'm not sure everyone needs to do the same, but I thought it might be of interest.

**Comment by ticky** [comment]
> To work around this issue _today_, I suggest just setting the `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` environment variable when working with Puma.

I’d warn that as discussed above, this potentially masks other issues which this new behaviour is attempting to mask. [Loading Foundation explicitly](https://github.com/puma/puma/issues/1421#issuecomment-332650703) will avoid the “implicit Obj-C runtime initialisation in a fork” errors, but _not_ mask similar but distinct errors in which your program actually triggers the behaviour this intends to disallow.

**Comment by nateberkopec** [comment]
> this potentially masks other issues which this new behaviour is attempting to mask

I think you meant "attempting to unmask" right?

My view is that Apple basically turned a feature into a bug here and made a helpful safety check cause a crash. Puma already warns when multiple threads exist before forking, so the safety check here is not as helpful as it might otherwise be. If people want the additional safety check Apple wants to provide here, they can upgrade to a newer Ruby version (when it comes out). Until then, they can live with the old behavior via the environment variable workaround.

**Comment by ticky** [comment]
> I think you meant "attempting to unmask" right?

Yep :)

> Puma already warns when multiple threads exist before forking, so the safety check here is not as helpful as it might otherwise be.

The safety check won’t apply to any C extensions which may fork, though, right?

**Comment by nateberkopec** [comment]
> The safety check won’t apply to any C extensions which may fork, though, right?

Correct, we're only checking the Ruby `Thread.list`.

**Comment by boazsegev** [comment]
> The safety check won’t apply to any C extensions which may fork, though, right?

@nateberkopec , I'm not sure what the common practice is, but I re-assesed the observation that language extensions (C / Java) could call `fork` directly.

It seems that Ruby has a "timer thread" running in the background. After reading through the Ruby source code I realized that calling `fork` directly might have negative effects on Ruby performance and scheduling.

Extensions that call `fork` directly might be considered buggy and could possibly be ignored as far as the safety check is concerned (on the grounds of these extensions leaving behind the Ruby scope).

As for the `Thread.list`, native threads don't work well with Ruby threads (due to the GIL and related data structures and signaling concerns). So the `Thread.list` could be considered complete as far as the safety check is concerned... with a single possible exception...

...however, some extensions might spawn new native threads that never call the Ruby API (iodine does this for IO flushing). These threads are obviously not in the `Thread.list`. But I think the safety warning can safely ignore these extensions, since they are clearly aware of these details and were designed with these considerations in mind.

**Comment by rhymes** [comment]
@nateberkopec @ticky unfortunately it seems the issue is still present even with Ruby 2.4.3

I just encountered it with Ruby 2.4.3, macOS High Sierra 10.13.2 and puma 3.11.0:

> [10996] Puma starting in cluster mode...
> [10996] * Version 3.11.0 (ruby 2.4.3-p205), codename: Love Song
> [10996] * Min threads: 1, max threads: 1
> [10996] * Environment: development
> [10996] * Process workers: 2
> [10996] * Preloading application
> [10996] * Listening on tcp://0.0.0.0:5000
> [10996] Use Ctrl-C to stop
> [10996] - Worker 0 (pid: 11016) booted, phase: 0
> [10996] - Worker 1 (pid: 11017) booted, phase: 0
> I, [2017-12-15T19:27:32.407998 #11016]  INFO -- : REQUEST: method=GET path=/api/v1/... user_agent=curl/7.57.0 time=2017-12-15T18:27:32Z ip=127.0.0.1 host=localhost params={}
> objc[11016]: +[__NSPlaceholderDictionary initialize] may have been in progress in another thread when fork() was called.
> objc[11016]: +[__NSPlaceholderDictionary initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
> [10996] - Worker 0 (pid: 11060) booted, phase: 0

I've isolated it, in my code, to these three lines of code which crash the process with our without preloading:

>     Faraday.new(url: url) do |faraday|
>       faraday.adapter :patron
>     end

@ticky patch did the trick

**Comment by webark** [comment]
this still seems to be happening for me in puma's cluster mode, not in single mode.
```
$ > bundle exec puma -w 3
[49550] Puma starting in cluster mode...
[49550] * Version 3.10.0 (ruby 2.4.3-p205), codename: Russell's Teapot
[49550] * Min threads: 0, max threads: 16
[49550] * Environment: development
[49550] * Process workers: 3
[49550] * Phased restart available
[49550] * Listening on tcp://0.0.0.0:9292
[49550] Use Ctrl-C to stop
objc[49556]: +[__NSPlaceholderDictionary initialize] may have been in progress in another thread when fork() was called.
```

**Comment by rhymes** [comment]
@webark yeah the issue is because of the usage of forking in OSX and how it changed in High Sierra, try @ticky's patch here: https://github.com/puma/puma/issues/1421#issue-260744683

**Comment by ticky** [comment]
@rhymes, @webark: the upstream Ruby patch hasn’t been back ported to Ruby 2.4. It looks like it will in fact land in Ruby 2.5, which is forthcoming.

@nateberkopec, it might be worth reconsidering the “no Puma workaround” approach given there will potentially be people wanting to stay on 2.4 and I don’t have confirmation of whether any backport is planned.

**Comment by ticky** [comment]
Update: [a backport for 2.4.x is forthcoming!](https://bugs.ruby-lang.org/issues/14009#note-21)

**Comment by rhymes** [comment]
@ticky amazing! thanks for the help and support!

**Comment by ticky** [comment]
Update: [a backport for 2.4.x has landed](https://bugs.ruby-lang.org/issues/14009#note-23), it should be in the 2.4.4 release, though I suspect that won’t appear until the new year!

**Comment by ndbroadbent** [comment]
Hi everyone! I am seeing the same `fork()` crash on a fresh install of macOS Catalina (10.15.1), and with `ruby 2.6.5p114` (installed with rbenv.) (I've also [opened an issue on the rbenv/ruby-build repo](https://github.com/rbenv/ruby-build/issues/1385).)

I first saw the crash when I ran `rails runner` with Spring. Then I found this GitHub issue, and I am seeing the same crash with `puma`:

```
$ rbenv install 2.6.5
Installing openssl-1.1.1d...
Installed openssl-1.1.1d to /Users/ndbroadbent/.rbenv/versions/2.6.5

Downloading ruby-2.6.5.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.6/ruby-2.6.5.tar.bz2
Installing ruby-2.6.5...
ruby-build: using readline from homebrew
Installed ruby-2.6.5 to /Users/ndbroadbent/.rbenv/versions/2.6.5

$ ruby --version
ruby 2.6.5p114 (2019-10-01 revision 67812) [x86_64-darwin19]

$ rails runner "User.first"
Running via Spring preloader in process 42270
objc[42270]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called.
objc[42270]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.

$ bundle exec puma -w 3
...
[43310] Use Ctrl-C to stop
[43310] - Worker 0 (pid: 43424) booted, phase: 0
[43310] - Worker 1 (pid: 43425) booted, phase: 0
[43310] - Worker 2 (pid: 43426) booted, phase: 0
Started GET "/" for 127.0.0.1 at 2019-11-30 06:05:02 +0700
objc[43424]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called.
objc[43424]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
```

I tested both `ruby 2.5.7p206` and `ruby 2.6.5p114`, and puma is crashing for both versions.

`export OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` prevents the crash, so I can add this to my `~/.bashrc` for now.

**Comment by boazsegev** [comment]
@ndbroadbent ,

I wish I could test this, but I run my music studio on the same computer and can't update to Catalina just yet (not until all the pro-audio software is compatible).

However, I thought this was solved in the Ruby core (this should probably be considered a Ruby bug rather than a Puma specific issue).

You can see the code that links agains the [Objective C Framework](https://www.evernote.com/shard/s199/nl/24966950/ac4c0e69-9f7d-4319-9e99-a76fe8f54ad9/) in the Ruby repo. It should have solved the issue... unless (maybe) you might be using custom compilation flags or a different compiler.

I am curious if a dynamic patch / workaround could be applied, [like I did with iodine](https://github.com/boazsegev/iodine/blob/50bf96b334b7c74c725198f621b38ca01e4efbc0/ext/iodine/iodine.c#L8-L23) (only using fiddle). Does the issue occur with iodine?

Kindly,
   Bo.

**Comment by dentarg** [comment]
@ndbroadbent can you open a new issue and describe how to reproduce the crash? (i.e. share code that makes it crash)

**Comment by kleisauke** [comment]
This should probably be re-opened since commit https://github.com/ruby/ruby/commit/8b182a7f7d798ab6539518fbfcb51c78549f9733 was implicitly reverted with commit https://github.com/ruby/ruby/commit/dc7044eb48f475e5ac34e994e89115052574c451. See for example the related Rails issue: https://github.com/rails/rails/issues/38560.

Note that an alternative workaround/solution would be to link Ruby on macOS with:
```
-Wl,-sectcreate,__DATA,__objc_fork_ok,/dev/null
```

That would be the same as setting the `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` env, but limited to the Ruby executable.
https://github.com/apple-oss-distributions/objc4/blob/objc4-906/runtime/objc-os.mm#L468-L482

**Comment by kleisauke** [comment]
Here's a standalone reproducer for this:
<details>
  <summary>Details</summary>

```c
// Simulate a child process initializing ObjC classes concurrently with another thread.
// Crashes unless the `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` environment variable is set,
// or the executable includes both `__DATA,__objc_fork_ok` and `__DATA_CONST,__objc_imageinfo` sections.
// Test with:
//   clang test.c -o test.out -framework CoreFoundation
//   OBJC_PRINT_INITIALIZE_METHODS=YES ./test.out
// And:
//   clang test.c -o test.out -framework CoreFoundation -Wl,-sectcreate,__DATA_CONST,__objc_imageinfo,/dev/null -Wl,-sectcreate,__DATA,__objc_fork_ok,/dev/null
//   OBJC_PRINT_INITIALIZE_METHODS=YES ./test.out

#include <errno.h>
#include <pthread.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

#include <CoreFoundation/CFString.h>

// Reference: https://github.com/ruby/ruby/blob/v3_4_2/file.c#L273-L324
static CFMutableStringRef mutable_CFString_new(CFStringRef *s, const char *ptr, long len) {
  const CFAllocatorRef alloc = kCFAllocatorDefault;
  *s = CFStringCreateWithBytesNoCopy(alloc, (const UInt8 *)ptr, len,
                                     kCFStringEncodingUTF8, FALSE,
                                     kCFAllocatorNull);
  return CFStringCreateMutableCopy(alloc, len, *s);
}

#define mutable_CFString_release(m, s) (CFRelease(m), CFRelease(s))

static void rb_CFString_class_initialize_before_fork(void) {
  const char small_str[] = "/";
  long len = sizeof(small_str) - 1;
  CFStringRef s;

  for (int i = 0; i < 2; i++) {
    CFMutableStringRef m = mutable_CFString_new(&s, small_str, len);
    mutable_CFString_release(m, s);
  }
}

void *thread_main(void *arg) {
  rb_CFString_class_initialize_before_fork();
  return NULL;
}

int main(void) {
  setbuf(stdout, NULL); // Disable buffering on stdout

  pthread_t thread;
  pthread_create(&thread, NULL, thread_main, NULL);
  pthread_detach(thread);

  pid_t pid = fork();
  if (pid > 0) {
    printf("PID %d (parent) waiting for child PID %d\n", getpid(), pid);

    waitpid(pid, NULL, 0);
  } else if (pid == 0) {
    printf("PID %d (child)\n", getpid());

    rb_CFString_class_initialize_before_fork();

    printf("Done!\n");
  } else {
    fprintf(stderr, "fork failed with: %s\n", strerror(errno));
    return -1;
  }

  return 0;
}
```
</details>

And a simple regression CLI test for ruby-vips, which currently fails:
```bash
$ ruby -e 'if pid = fork
  p Process.wait2(pid)
else
  require "vips"
end'
```

**Comment by kleisauke** [comment]
To workaround this issue when using `rbenv`, use:
```console
$ rbenv install 3.4.4 -- LDFLAGS="-Wl,-sectcreate,__DATA_CONST,__objc_imageinfo,/dev/null -Wl,-sectcreate,__DATA,__objc_fork_ok,/dev/null"
$ rbenv local 3.4.4
$ ruby -e 'if pid = fork; p Process.wait2(pid) else require "vips"; end'
[17702, #<Process::Status: pid 17702 exit 0>]
```

**Comment by kliuless** [comment]
This specific comment on the rails thread explains the revert:
https://github.com/rails/rails/issues/38560#issuecomment-1881848901

> This should probably be re-opened since commit [ruby/ruby@8b182a7](https://github.com/ruby/ruby/commit/8b182a7f7d798ab6539518fbfcb51c78549f9733) was implicitly reverted with commit [ruby/ruby@dc7044e](https://github.com/ruby/ruby/commit/dc7044eb48f475e5ac34e994e89115052574c451). See for example the related Rails issue: [rails/rails#38560](https://github.com/rails/rails/issues/38560).
> 
> Note that an alternative workaround/solution would be to link Ruby on macOS with:
> 
> ```
> -Wl,-sectcreate,__DATA,__objc_fork_ok,/dev/null
> ```
> 
> That would be the same as setting the `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` env, but limited to the Ruby executable. https://github.com/apple-oss-distributions/objc4/blob/objc4-906/runtime/objc-os.mm#L468-L482

**Comment by kliuless** [comment]
While we wait for a fix from the ruby core team, this is a possible way to leverage the Gemfile to avoid repeating gem names in `before_fork` code:

Gemfile:
```ruby
# if the gem wasn't in any groups yet, ensure it's still in `:default`
gem 'fiddle', groups: [:default, :before_fork]

# if the gem was already in a `group` block
group :abc do
  gem 'fiddle', groups: [:before_fork]
end
```

In `before_fork` or equivalent:
```ruby
Bundler.definition.dependencies_for([:before_fork]).map(&:name).each { |gem| require gem }
```
